File: app/src/main/java/hottopic/mit/co/nz/cleaningservice/view/home/HomeActivity.java
package hottopic.mit.co.nz.cleaningservice.view.home;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Handler;
import android.os.Message;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.lsjwzh.widget.recyclerviewpager.RecyclerViewPager;

import java.util.ArrayList;
import java.util.List;

import hottopic.mit.co.nz.cleaningservice.BaseActivity;
import hottopic.mit.co.nz.cleaningservice.Constants;
import hottopic.mit.co.nz.cleaningservice.R;
import hottopic.mit.co.nz.cleaningservice.adapter.ServiceAdapter;
import hottopic.mit.co.nz.cleaningservice.adapter.ViewPagerAdapter;
import hottopic.mit.co.nz.cleaningservice.entities.orders.MainServiceType;
import hottopic.mit.co.nz.cleaningservice.entities.orders.ServiceTypes;
import hottopic.mit.co.nz.cleaningservice.entities.orders.SubServiceType;
import hottopic.mit.co.nz.cleaningservice.entities.orders.TitleModel;
import hottopic.mit.co.nz.cleaningservice.presenter.home.HomePresenterImpl;
import hottopic.mit.co.nz.cleaningservice.view.order.OrderBookingActivity;
import hottopic.mit.co.nz.cleaningservice.view.order.OrdersActivity;
import hottopic.mit.co.nz.cleaningservice.view.user.LoginActivity;

public class HomeActivity extends BaseActivity implements IHomeView, RecyclerViewPager.OnPageChangedListener, ServiceAdapter.OnItemClickedListener {
    private TextView tv_title, tv_page_number;
    private RecyclerView rv_cleaning;
    private ServiceAdapter cleaningAdapter;
    private ImageView iv_icon;
    private RelativeLayout lyt_right;
    private RecyclerViewPager rvp_ad;
    private ServiceTypes serviceTypes;
    private HomePresenterImpl homePresenter;

    private int totalPages = 1;
    private int autoPage = 0;
    private boolean isQuit = false;

    @SuppressLint("HandlerLeak")
    private Handler handler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            switch (msg.what){
                case Constants.WHAT_ADVERTISEMENT:
                    autoPage++;
                    if (autoPage > totalPages) {
                        autoPage = 0;
                        rvp_ad.scrollToPosition(autoPage);
                    } else {
                        rvp_ad.scrollToPosition(autoPage);
                    }
                    handler.sendEmptyMessageDelayed(Constants.WHAT_ADVERTISEMENT, 3000);
                    break;
                case Constants.WHAT_EXIT:
                    isQuit = false;
                    break;
            }
        }
    };

    private static final int[] ad_list = {R.drawable.advertisement_1, R.drawable.advertisement_2, R.drawable.advertisement_3,
            R.drawable.advertisement_4};

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        int statusType = intent.getIntExtra(Constants.KEY_INTENT_CLOSETYPE, 0);
        if (Constants.CLOSETYPE_LOGOUT == statusType){
            //open login page
            Intent login = new Intent(this, LoginActivity.class);
            startActivity(login);
            finish();
        }
    }

    @Override
    public void initView() {
        setContentView(R.layout.activity_home);
        tv_title = findViewById(R.id.tv_title);
        tv_page_number = findViewById(R.id.tv_page_number);
        rv_cleaning = findViewById(R.id.rv_cleaning);
        rvp_ad = findViewById(R.id.rvp_ad);
        lyt_right = findViewById(R.id.lyt_right);
        iv_icon = findViewById(R.id.iv_icon);
    }

    @Override
    public void initData() {
        tv_title.setText("HOME");
        lyt_right.setVisibility(View.VISIBLE);
        iv_icon.setImageResource(R.drawable.icon_user);
        serviceTypes = (ServiceTypes) getIntent().getSerializableExtra(Constants.KEY_INTENT_SERVICETYPE);
        homePresenter = new HomePresenterImpl(this, this);
        if (serviceTypes == null){
            homePresenter.getServiceTypes();
        }else{
            viewDisplay(serviceTypes);
        }
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);//LinearLayoutManager.HORIZONTAL 设置水平滚动
        rvp_ad.setLayoutManager(layoutManager);
        rvp_ad.setAdapter(new ViewPagerAdapter(this, ad_list));
        totalPages = ad_list.length;
        tv_page_number.setText("1/" + totalPages);
        //Start the automatic page turning
        handler.sendEmptyMessageDelayed(Constants.WHAT_ADVERTISEMENT, 3000);
    }

    @Override
    public void initListener() {
        rvp_ad.addOnPageChangedListener(this);
        cleaningAdapter.setOnItemClickedListener(this);
        lyt_right.setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()){
            case R.id.lyt_right:
                Intent intent = new Intent(HomeActivity.this, OrdersActivity.class);
                startActivity(intent);
                break;
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            //If click the Back key twice in 2 seconds, the program exits
            if (!isQuit) {
                isQuit = true;
                Toast.makeText(getApplicationContext(), "Press the return key again to exit the program!",
                        Toast.LENGTH_SHORT).show();
                handler.sendEmptyMessageDelayed(Constants.WHAT_EXIT, 2000);
            } else {
                finish();
                System.exit(0);
            }
        }
        return false;
    }

    @Override
    public void getSerivceError(String message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void getServiceTypes(ServiceTypes serviceTypes) {
        viewDisplay(serviceTypes);
    }

    private void viewDisplay(ServiceTypes serviceTypes){
        List dataList = new ArrayList();
        List<MainServiceType> mainServiceTypes = serviceTypes.getMainServiceTypes();
        for (MainServiceType mainServiceType : mainServiceTypes){
            dataList.add(new TitleModel(mainServiceType.getType_name()));
            dataList.addAll(mainServiceType.getSubServiceTypes());
        }
        initServiceAdapter(dataList);
    }

    private void initServiceAdapter(List dataList) {
        cleaningAdapter = new ServiceAdapter(this);
        cleaningAdapter.setData(dataList);
        rv_cleaning.setAdapter(cleaningAdapter);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(this, 2);
        gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                if (dataList.get(position) instanceof TitleModel) {
                    return 2;
                } else {
                    return 1;
                }
            }
        });
        rv_cleaning.setLayoutManager(gridLayoutManager);
        rv_cleaning.setHasFixedSize(true);
        rv_cleaning.setNestedScrollingEnabled(false);
        rv_cleaning.setFocusable(false);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void OnPageChanged(int oldPosition, int newPosition) {
        //User to show the current photo position
        Log.d("OnPageChanged", newPosition + "");
        tv_page_number.setText(newPosition + 1 + "/" + totalPages);
        autoPage = newPosition;
    }

    @Override
    public void OnItemClicked(SubServiceType serviceType) {
        Intent intent = new Intent(HomeActivity.this, OrderBookingActivity.class);
        intent.putExtra(Constants.KEY_INTENT_SERVICETYPE, serviceType);
        startActivity(intent);
    }
}

File: app/src/main/java/hottopic/mit/co/nz/cleaningservice/view/order/implementation/HomeActivity.kt
package hottopic.mit.co.nz.cleaningservice.view.order.implementation

import android.annotation.SuppressLint
import android.content.Intent
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.widget.Toast
import com.lsjwzh.widget.recyclerviewpager.RecyclerViewPager
import hottopic.mit.co.nz.cleaningservice.BaseActivity
import hottopic.mit.co.nz.cleaningservice.Constants
import hottopic.mit.co.nz.cleaningservice.R
import hottopic.mit.co.nz.cleaningservice.adapter.ServiceAdapter
import hottopic.mit.co.nz.cleaningservice.adapter.ViewPagerAdapter
import hottopic.mit.co.nz.cleaningservice.entities.orders.ServiceTypes
import hottopic.mit.co.nz.cleaningservice.entities.orders.SubServiceType
import hottopic.mit.co.nz.cleaningservice.entities.orders.TitleModel
import hottopic.mit.co.nz.cleaningservice.presenter.order.implementation.OrderPresenterImpl
import hottopic.mit.co.nz.cleaningservice.view.order.inter.IHomeView
import hottopic.mit.co.nz.cleaningservice.view.user.implementation.LoginActivity
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_home.*
import kotlinx.android.synthetic.main.view_title_bar.*
import java.util.concurrent.TimeUnit

class HomeActivity : BaseActivity(), IHomeView, RecyclerViewPager.OnPageChangedListener, ServiceAdapter.OnItemClickedListener {
    companion object {
        private val ad_list = intArrayOf(R.drawable.advertisement_1, R.drawable.advertisement_2, R.drawable.advertisement_3, R.drawable.advertisement_4)
    }
    private lateinit var cleaningAdapter: ServiceAdapter
    private lateinit var mSubscription: CompositeDisposable

    private var totalPages = 1
    private var autoPage = 0
    private var isQuit = false

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        val statusType = intent.getIntExtra(Constants.KEY_INTENT_CLOSETYPE, 0)
        if (Constants.CLOSE_TYPE_LOGOUT == statusType) {
            //open login page
            val login = Intent(this, LoginActivity::class.java)
            startActivity(login)
            finish()
        }
    }

    override fun initView() {
        setContentView(R.layout.activity_home)

    }

    @SuppressLint("SetTextI18n")
    override fun initData() {
        tvTitle.text = "HOME"
        lytRight.visibility = View.VISIBLE
        ivIcon.setImageResource(R.drawable.icon_purchase)
        val serviceTypes = intent.getSerializableExtra(Constants.KEY_INTENT_SERVICETYPE) as ServiceTypes?
        mSubscription = CompositeDisposable()
        cleaningAdapter = ServiceAdapter(this)
        if(serviceTypes != null) viewDisplay(serviceTypes)
        initAdvert()
    }

    override fun initListener() {
        rvp_ad.addOnPageChangedListener(this)
        lytRight.setOnClickListener(this)
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.lytRight -> navigateToOrder()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mSubscription.clear()
    }

    @SuppressLint("CheckResult")
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (keyCode == KeyEvent.KEYCODE_BACK) {
                //If click the Back key twice in 2 seconds, the program exits
                if (!isQuit) {
                    isQuit = true
                    Toast.makeText(applicationContext, "Press the return key again to exit the program!",
                            Toast.LENGTH_SHORT).show()
                    Observable.timer(Constants.EXIT_WAIT_TIME, TimeUnit.SECONDS)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe{isQuit = false}
                } else {
                    finish()
                    System.exit(0)
                }
            }
        }
        return false
    }

    override fun navigateToOrder() {
        val intent = Intent(this@HomeActivity, OrdersActivity::class.java)
        startActivity(intent)
    }

    override fun goToAdverse() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun navigateToBooking(serviceType: SubServiceType) {
        val intent = Intent(this@HomeActivity, OrderBookingActivity::class.java)
        intent.putExtra(Constants.KEY_INTENT_SERVICETYPE, serviceType)
        startActivity(intent)
    }

    private fun viewDisplay(serviceTypes: ServiceTypes) {
        val dataList = mutableListOf<Any>()
        val mainServiceTypes = serviceTypes.mainServiceTypes
        if (mainServiceTypes != null) {
            for (mainServiceType in mainServiceTypes) {
                dataList.add(TitleModel(mainServiceType.type_name!!))
                dataList.addAll(mainServiceType.subServiceTypes!!)
            }
        }
        initServiceAdapter(dataList)
    }

    @SuppressLint("CheckResult", "SetTextI18n")
    private fun initAdvert(){
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        rvp_ad.layoutManager = layoutManager
        val adapter = ViewPagerAdapter(this)
        adapter.setData(ad_list)
        rvp_ad.adapter = adapter
        totalPages = ad_list.size
        tv_page_number.text = "1/$totalPages"

        val subscription = Observable.interval(3, 3, TimeUnit.SECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe {
                    autoPage++
                    if (autoPage > totalPages) {
                        autoPage = 0
                        rvp_ad!!.scrollToPosition(autoPage)
                    } else {
                        rvp_ad!!.scrollToPosition(autoPage)
                    }
                }
        mSubscription.add(subscription)
    }
    private fun initServiceAdapter(dataList: List<*>) {
        cleaningAdapter.setData(dataList)
        cleaningAdapter.setOnItemClickedListener(this)
        rv_cleaning.adapter = cleaningAdapter
        val gridLayoutManager = GridLayoutManager(this, 2)
        gridLayoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                return if (dataList[position] is TitleModel) {
                    2
                } else {
                    1
                }
            }
        }
        rv_cleaning.layoutManager = gridLayoutManager
        rv_cleaning.setHasFixedSize(true)
        rv_cleaning.isNestedScrollingEnabled = false
        rv_cleaning.isFocusable = false
    }

    @SuppressLint("SetTextI18n")
    override fun OnPageChanged(oldPosition: Int, newPosition: Int) {
        //User to show the current photo position
        Log.d("OnPageChanged", newPosition.toString() + "")
        tv_page_number!!.text = (newPosition + 1).toString() + "/" + totalPages
        autoPage = newPosition
    }

    override fun onItemClicked(serviceType: SubServiceType?) {
        if (serviceType != null) {
            navigateToBooking(serviceType)
        }
    }
}
